### PR TITLE
[EHN] Including EEGDeformer

### PR DESCRIPTION
**Pull request by bruAristimunha**

### CODE REVIEW

**Comment by martinwimpff** [comment]
Is there any reason to not use torchs [TransformerEncoderLayer](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html#torch.nn.TransformerEncoderLayer) instead of a custom `FeedForward `+ `Attention `module?

File: braindecode/models/eegdeformer.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    ):
+        super().__init__()
+        self.layers = nn.ModuleList([])
+        for i in range(depth):
+            dim = int(dim * 0.5)
+            self.layers.append(
+                nn.ModuleList(
+                    [
+                        _Attention(dim, heads, dim_head, dropout=dropout),
+                        _FeedForward(dim, mlp_dim, drop_prob=dropout),
```
Comment (martinwimpff) [inline braindecode/models/eegdeformer.py:135]:
```suggestion
        embed_dim,
        depth,
        heads,
        dim_head,
        dim_feedforward,
        in_chan,
        fine_grained_kernel=11,
        dropout=0.0,
    ):
        super().__init__()
        self.layers = nn.ModuleList([])
        for i in range(depth):
            embed_dim= int(embed_dim* 0.5)
            self.layers.append(
                nn.ModuleList(
                    [
                        _Attention(embed_dim, heads, dim_head, dropout=dropout),
                        _FeedForward(embed_dim, dim_feedforward, drop_prob=dropout),
```
consistent naming in itself and with torch

File: braindecode/models/eegdeformer.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+       Interfaces. arXiv preprint arXiv:2405.00719.
+    """
+
+    def __init__(
+        self,
+        temporal_kernel: int = 11,
+        num_kernel: int = 64,
+        n_layers: int = 4,
+        heads: int = 16,
+        mlp_dim: int = 16,
```
Comment (martinwimpff) [inline braindecode/models/eegdeformer.py:211]:
```suggestion
        dim_feedforward: int = 16,
```

File: braindecode/models/eegdeformer.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        # Variables
+        self.n_layers = n_layers
+        self.drop_prob = drop_prob
+        self.dim = int(0.5 * self.n_times)
+        self.hidden_size = int(num_kernel * int(self.dim * (0.5**n_layers))) + int(
+            num_kernel * n_layers
+        )
+
+        # Parameters
+        self.pos_embedding = nn.Parameter(torch.randn(1, num_kernel, self.dim))
```
Comment (martinwimpff) [inline braindecode/models/eegdeformer.py:241]:
```suggestion
        self.embed_dim= int(0.5 * self.n_times)
        self.hidden_size = int(num_kernel * int(self.embed_dim* (0.5**n_layers))) + int(
            num_kernel * n_layers
        )

        # Parameters
        self.pos_embedding = nn.Parameter(torch.randn(1, num_kernel, self.embed_dim))
```

File: braindecode/models/eegdeformer.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+        self.to_patch_embedding = Rearrange(
+            "batch kernel chans filter -> batch kernel (chans filter)"
+        )
+        self.transformer = _Transformer(
+            dim=self.dim,
+            depth=self.n_layers,
+            heads=heads,
+            dim_head=dim_head,
+            mlp_dim=mlp_dim,
```
Comment (martinwimpff) [inline braindecode/models/eegdeformer.py:274]:
```suggestion
            embed_dim=self.embed_dim,
            depth=self.n_layers,
            heads=heads,
            dim_head=dim_head,
            dim_feedforward=dim_feedforward,
```

File: braindecode/models/eegdeformer.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+        # Layers
+        self.ensuredim = Rearrange("batch chan time -> batch 1 chan time")
+
+        self.cnn_encoder = nn.Sequential(
+            Conv2dWithConstraint(
+                in_channels=1,
+                out_channels=num_kernel,
+                kernel_size=(1, temporal_kernel),
+                padding=(0, int(0.5 * (temporal_kernel - 1))),
```
Comment (martinwimpff) [inline braindecode/models/eegdeformer.py:251]: Is this a bug or a feature? I would either use `temporal_kernel//2` or `"same"` but this would change the padding for even kernel_lengths.

Comment (bruAristimunha) [reply braindecode/models/eegdeformer.py]: bug, I solved in the last commit

File: braindecode/models/eegdeformer.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
+"""
+CBCR License 1.0
+
+Copyright 2022 Centre for Brain Computing Research (CBCR)
+
```
Comment (agramfort) [inline braindecode/models/eegdeformer.py:5]: I don’t think the license is compatible with our bsd license

Comment (bruAristimunha) [reply braindecode/models/eegdeformer.py:5]: Hmm, okay, thank you for catching this. I considered attributing only this code to this particular license, but I think it would not be possible. I was benchmarking the model a little, and the performance is not super good for BCI

Comment (martinwimpff) [reply braindecode/models/eegdeformer.py:5]: @bruAristimunha can you share the (preliminary) results?
